README.md
[![Build Status](https://github.com/shinesolutions/aem-stack-manager-messenger/workflows/CI/badge.svg)](https://github.com/shinesolutions/aem-stack-manager-messenger/actions?query=workflow%3ACI)
[![Known Vulnerabilities](https://snyk.io/test/github/shinesolutions/aem-stack-manager-messenger/badge.svg)](https://snyk.io/test/github/shinesolutions/aem-stack-manager-messenger)

AEM Stack Manager Messenger
---------------------------

A set of [Ansible](https://www.ansible.com/) playbooks for triggering events on [AEM Stack Manager](https://github.com/shinesolutions/aem-stack-manager-cloud) environment which is built using [AEM AWS Stack Builder](https://github.com/shinesolutions/aem-aws-stack-builder).

Learn more about AEM Stack Manager Messenger:

* [Installation](https://github.com/shinesolutions/aem-stack-manager-messenger#installation)
* [Configuration](https://github.com/shinesolutions/aem-stack-manager-messenger/blob/master/docs/configuration.md)
* [Usage](https://github.com/shinesolutions/aem-stack-manager-messenger#usage)
* [Testing](https://github.com/shinesolutions/aem-stack-manager-messenger#testing)
* [Frequently Asked Questions](https://github.com/shinesolutions/aem-stack-manager-messenger/blob/master/docs/faq.md)

AEM Stack Manager Messenger is part of [AEM OpenCloud](https://aemopencloud.io) platform.

Installation
------------

- Either clone AEM Stack Manager Messenger `git clone https://github.com/shinesolutions/aem-stack-manager-messenger.git` or download one of the [released versions](https://github.com/shinesolutions/aem-stack-manager-messenger/releases)
- Install the following required tools:
  * [Python](https://www.python.org/downloads/) version 2.7.x
  * [GNU Make](https://www.gnu.org/software/make/)<br/>

  Alternatively, you can use [AEM Platform BuildEnv](https://github.com/shinesolutions/aem-platform-buildenv) Docker container to run AEM Stack Manager Messenger build targets.
- Resolve the [Python packages](https://github.com/shinesolutions/aem-stack-manager-messenger/blob/master/requirements.txt) dependencies by running `make deps`

Usage
-----

- Create [configuration file](https://github.com/shinesolutions/aem-stack-manager-messenger/blob/master/docs/configuration.md)
- Execute the AEM Stack Manager events:

Deploy a set of AEM Packages and Dispatcher artifacts into an AEM environment using a [Deployment Descriptor](https://github.com/shinesolutions/aem-aws-stack-builder/blob/master/docs/descriptors.md#deployment-descriptor) file:

    make deploy-artifacts-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir> \
      descriptor_file=deploy-artifacts-descriptor.json

    make deploy-artifacts-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir> \
      descriptor_file=deploy-artifacts-descriptor.json

Deploy a single AEM Package into an AEM instance:

    make deploy-artifact \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      aem_id=author \
      source=s3://s3bucket/path/aem-helloworld-content-0.0.1-SNAPSHOT.zip \
      group=shinesolutions \
      name=aem-helloworld-content \
      version=0.0.1 \
      replicate=true \
      activate=false \
      force=true

Disable SAML on an AEM instance:

    make disable-saml \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      aem_id=author path='/'

Enable SAML with filepath to SAML certificate on an AEM instance:

    make enable-saml \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      aem_id=author path='/' \
      file="s3://aem-opencloud/adobeaemcloud/certs/saml_cert.pem" \
      assertion_consumer_service_url="https://author.aemopencloud.net:443/saml_login" \
      service_provider_entity_id=AEMSSO \
      idp_url="https://federation.server.com" \
      create_user=true \
      default_groups=content-authors \
      group_membership_attribute=groupMembership \
      handle_logout=true \
      idp_http_redirect=false \
      logout_url="https://accounts.google.com/logout" \
      name_id_format='urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress' \
      signature_method="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" \
      synchronize_attributes="givenName=profile/givenName,familyName=profile/familyName,mail=profile/email" \
      use_encryption=false \
      user_id_attribute=NameID

Enable SAML with IDP Certalias of the SAML certificate as in the AEM Truststore on an AEM instance:

    make enable-saml \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      aem_id=author path='/' \
      idp_cert_alias='certalias_1234' \
      assertion_consumer_service_url="https://author.aemopencloud.net:443/saml_login" \
      service_provider_entity_id=AEMSSO \
      idp_url="https://federation.server.com" \
      create_user=true \
      default_groups=content-authors \
      group_membership_attribute=groupMembership \
      handle_logout=true \
      idp_http_redirect=false \
      logout_url="https://accounts.google.com/logout" \
      name_id_format='urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress' \
      signature_method="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" \
      synchronize_attributes="givenName=profile/givenName,familyName=profile/familyName,mail=profile/email" \
      use_encryption=false \
      user_id_attribute=NameID

Enable SAML with defined SERIAL number of the SAML certificate on an AEM instance:

    make enable-saml \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      aem_id=author path='/' \
      serial='1234567890' \
      assertion_consumer_service_url="https://author.aemopencloud.net:443/saml_login" \
      service_provider_entity_id=AEMSSO \
      idp_url="https://federation.server.com" \
      create_user=true \
      default_groups=content-authors \
      group_membership_attribute=groupMembership \
      handle_logout=true \
      idp_http_redirect=false \
      logout_url="https://accounts.google.com/logout" \
      name_id_format='urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress' \
      signature_method="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" \
      synchronize_attributes="givenName=profile/givenName,familyName=profile/familyName,mail=profile/email" \
      use_encryption=false \
      user_id_attribute=NameID

Enable CRXDE on an AEM instance:

    make enable-crxde \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary

Disable CRXDE on an AEM instance:

    make disable-crxde \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary

Export a set of backup packages from an AEM instance using a [Package Backup Descriptor](https://github.com/shinesolutions/aem-aws-stack-builder/blob/master/docs/descriptors.md#package-backup-descriptor) file:

    make export-packages-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir> \
      descriptor_file=<package_backup_descriptor_file>

    make export-packages-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir> \
      descriptor_file=<package_backup_descriptor_file>

Export a single AEM package from an AEM instance:

    make export-package \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      package_group=somegroup \
      package_name=somepackage \
      package_filter='[{"root":"/apps/geometrixx","rules":[]},{"root":"/apps/geometrixx-common","rules":[]}]'

Flush dispatcher cache:

    make flush-dispatcher-cache \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=publish-dispatcher

Import a single AEM package into an AEM instance:

    make import-package \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary \
      source_stack_prefix=stack2 \
      package_group=somegroup \
      package_name=somepackage \
      package_datestamp=201702

List AEM packages available on an AEM instance:

    make list-packages \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary

Promote an AEM Author Standby to become AEM Author Primary:

    make promote-author \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Take live snapshot of the repositories within an AEM component:

    make live-snapshot \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-primary

Take offline snapshots of a single AEM Author and a single AEM Publish repositories within an AEM environment (See offline-snapshot-full-set):

    make offline-snapshot \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Take offline snapshots of a single AEM Author and a single AEM Publish repositories within an AEM environment and run offline compaction on those AEM instances (See offline-compaction-snapshot-full-set):

    make offline-compaction-snapshot \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Take offline snapshots of a single AEM Author and a single AEM Publish repositories within an AEM environment (currently only supports Full-Set):

    make offline-snapshot-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Take offline snapshots of a single AEM Author and a single AEM Publish repositories within an AEM environment and run offline compaction on those AEM instances (currently only supports Full-Set):

    make offline-compaction-snapshot-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Take offline snapshots of a single AEM Author and a single AEM Publish repositories within an AEM Consolidated instance:

    make offline-snapshot-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Take offline snapshots of a single AEM Author and a single AEM Publish repositories within an AEM Consolidated instance and run offline compaction on the AEM instance:

    make offline-compaction-snapshot-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_stack_prefix> \
      config_path=<path/to/config/dir>

Check readiness of an AEM environment:

    make check-readiness-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir>

    make check-readiness-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir>

Schedule/unschedule jobs on an AEM environment:

    make schedule-offline-snapshot-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir>

    make unschedule-offline-snapshot-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir>

    make schedule-offline-compaction-snapshot-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir>

    make unschedule-offline-compaction-snapshot-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix> \
      config_path=<path/to/config/dir>

    make schedule-offline-snapshot-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir>

    make unschedule-offline-snapshot-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir>

    make schedule-offline-compaction-snapshot-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir>

    make unschedule-offline-compaction-snapshot-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir>

Upgrade AEM environment to a new version on AEM environment:
    AEM Upgrade preparation step 1 download AEM e.g. AEM 6.4 jar file and unpack it

    make upgrade-unpack-jar \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-publish-dispatcher \
      aem_artifacts_base="s3://aem-opencloud/adobeaemcloud" \
      aem_upgrade_version="6.4" \
      enable_backup=false

    AEM Upgrade preparation step 2 run repository migration

    make upgrade-unpack-jar \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-publish-dispatcher \
      source_crx2oak="https://repo.adobe.com/nexus/content/groups/public/com/adobe/granite/crx2oak/1.8.6/crx2oak-1.8.6-all-in-one.jar"

    Trigger the AEM Upgrade to specified version:

    make upgrade-unpack-jar \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix> \
      config_path=<path/to/config/dir> \
      component=author-publish-dispatcher \
       aem_upgrade_version="6.4"

Testing
-------

You can run integration test for running all AEM Stack Manager events for a given AEM architecture, using the latest [AEM Hello World Config](https://github.com/shinesolutions/aem-helloworld-config) from GitHub.

Run integration tests for AEM Full-Set architecture:

    make test-full-set \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix>

Run integration tests for AEM Consolidated architecture:

    make test-consolidated \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix>

Alternatively, you can also run the integration tests using a local AEM Hello World Config.

    make test-full-set-local \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_full_set_stack_prefix>

    make test-consolidated-local \
      stack_prefix=<stack_manager_stack_prefix> \
      target_aem_stack_prefix=<aem_consolidated_stack_prefix>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 2.15.2 - 2022-08-23
### Changed
- Revert the fix cloudformation_facts no longer returning any result [#89]

## 2.15.1 - 2022-08-17
### Fixed
- Fix cloudformation_facts no longer returning any result [#89]

## 2.15.0 - 2021-11-10
### Changed
- Update awscli to 1.22.1
- Update boto3 to 1.20.0
- Update Ansible to 4.8.0

## 2.14.2 - 2021-10-06
### Added
- Add release-major, release-minor, release-patch, and publish Makefile targets and GitHub Actions

## 2.14.1 - 2021-03-12
### Changed
- Update Ansible to 3.1.0

## 2.14.0 - 2021-03-02
### Changed
- Update Ansible to 3.0.0
- Update awscli to 1.19.8
- update boto3 to 1.17.8

### Fixed
- Fixed hosts to use python3 instead of python

## 2.13.0 - 2021-02-11
### Changed
- Lock down pylint to 2.6.0
- Use pip3 for python package management
- Convert python scripts to be executed using Python 3
- Set default Ansible Python interpreter to python3

## 2.12.0 - 2020-04-27
### Changed
- Update default executionTimeout for package deployment, export & import commands [shinesolutions/aem-aws-stack-builder#401]

## 2.11.0 - 2020-02-28
### Added
- Add new ssm parameter 'executionTimeout'

### Changed
- Changed offline-snapshot & offline-compatcion-snapshot default retries. It will now use the retries configured in the configuration profile

## 2.10.0 - 2020-01-20
### Changed
- Change integration test helloworld content package to use https://repo.maven.apache.org

## 2.9.0 - 2019-11-12
### Added
- Add feature to trigger Toughday2 performance test

### Changed
- Improved conditional checks when querying/scanning the DynamoDB [#77]

## 2.8.0 - 2019-08-20
### Added
- Add enable-saml and disable-saml

## 2.7.0 - 2019-08-14
### Changed
- Changed test-readiness-full-set template to send it only component orchestrator [#75]

### Removed
- Removed Make target for check-readiness-full-set-with-disabled-chaosmonkey [#75]
- Removed sns template check-readiness-full-set-with-disabled-chaosmonkey.json [#75]

## 2.6.0 - 2019-07-31
### Fixed
- Fixed issue with missing stage dir for sns message template file generation [#70]

## 2.5.0 - 2019-07-27
### Fixed
- Fixed offline-snapshot message sending sequence [#70]

## 2.4.2 - 2019-07-26
### Fixed
- Fixed offline-snapshot bug with resolving JSON template variables following upgrade to Ansible 2.8.x [#70]

## 2.4.1 - 2019-07-26
### Fixed
- Fixed bug with resolving JSON template variables following upgrade to Ansible 2.8.x [#70]

## 2.4.0 - 2019-07-21
### Added
- Add new make target to check stack readiness for stacks with disabled ChaosMonkey instance

## 2.3.1 - 2019-05-22
### Fixed
- Fix stack_prefix, target_aem_stack_prefix, and config_path variables assignment

## 2.3.0 - 2019-05-22
### Fixed
- Fix config_path variable assignment in run-playbook.sh script

## 2.2.1 - 2019-03-13
### Fixed
- Fix config path in run-playbook.sh script [#64]

## 2.2.0 - 2019-02-04
### Fixed
- Fix integration test config path since the introduction of profiles in aem-helloworld-config

## 2.1.0 - 2019-01-31
### Added
- Add Stack Manager events for AEM Upgrade automation.
- Add YAML checks to lint target

## 2.0.0 - 2018-12-07
### Changed
- Set test-readiness-full-set to run on all Full-Set architecture components

## 1.5.7 - 2018-11-05
### Added
- Add new Stack manager event to un/schedule live snapshot shinesolutions/aem-aws-stack-builder#212

## 1.5.6 - 2018-10-09
### Changed
- Fix SNS result message ID retrieval

## 1.5.5 - 2018-10-09
### Changed
- Fix missing region on AEM Stack existence check [#29]

## 1.5.4 - 2018-10-06
### Changed
- Change default logging to always dump log

## 1.5.3 - 2018-10-06
### Added
- Add feature to check if the target AEM Stack exists [#29]

### Changed
- Temporarily add unreleased ansible SNS module upgrade to boto3. [#59]

## 1.5.2 - 2018-09-11
### Removed
- Move all unschedule jobs to run at the beginning of the tests, all schedule jobs to run at the end

## 1.5.1 - 2018-08-19
### Added
- Add export-package & import-package to integration test

## 1.5.0 - 2018-07-31
### Added
- Add offline-snapshot, offline-compaction-snapshot for AEM Consolidated
- Add new Stack Manager event install-aem-profile

### Changed
- Rename offline-snapshot, offline-compaction-snapshot with full-set suffix
- Replace all check_message_sent usage with check_command_execution

### Removed
- Remove poll_timeout.check_message_sent.retries and poll_timeout.check_message_sent.delay configurations

## 1.4.1 - 2018-07-08
### Added
- Adding external id for SSM command execution

### Changed
- Improved error handling for SSM command execution

## 1.4.0 - 2018-05-24
### Changed
- Parameterise Stack Manager's main stack name [#44]
- Increase check command execution timeout to retry 720 times with 10 secs delay

## 1.3.1 - 2018-05-19
### Changed
- Check readiness no longer retries on failure now that Stack Manager event's SSM command ensures cloud-init completeness [#45]

## 1.3.0 - 2018-05-11
### Added
- Add message_config_file for author-publish-dispatcher
- Add poll_timeout.check_command_execution.retries and poll_timeout.check_command_execution.delay configurations [#31]
- Add command test-readiness-full-set
- Add command test-readiness-consolidated
- Add aem_id parameter to deploy-artifact and export-package message payloads
- Add poll_timeout.check_message_sent.retries and poll_timeout.check_message_sent.delay configurations
- Add show_log_on_failure_only configuration [#32]
- Add schedule/unschedule offline snapshot support

### Changed
- Replace topic_config parameter with config_path, which accepts a directory containing config files [#7]
- Fix export package message payload

ansible/library/sns.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# THIS FILE SHOULD BE REMOVED WHEN github.com/ansible/ansible/pull/45634 IS MERGED AND RELEASED

# (c) 2014, Michael J. Schultz <mjschultz@gmail.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type


ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}


DOCUMENTATION = """
module: sns
short_description: Send Amazon Simple Notification Service messages
description:
  - Sends a notification to a topic on your Amazon SNS account.
version_added: 1.6
author:
  - Michael J. Schultz (@mjschultz)
  - Paul Arthur (@flowerysong)
options:
  msg:
    description:
      - Default message for subscriptions without a more specific message.
    required: true
    aliases: [ "default" ]
  subject:
    description:
      - Message subject
  topic:
    description:
      - The name or ARN of the topic to publish to.
    required: true
  email:
    description:
      - Message to send to email subscriptions.
  email_json:
    description:
      - Message to send to email-json subscriptions
    version_added: '2.8'
  sqs:
    description:
      - Message to send to SQS subscriptions
  sms:
    description:
      - Message to send to SMS subscriptions
  http:
    description:
      - Message to send to HTTP subscriptions
  https:
    description:
      - Message to send to HTTPS subscriptions
  application:
    description:
      - Message to send to application subscriptions
    version_added: '2.8'
  lambda:
    description:
      - Message to send to Lambda subscriptions
    version_added: '2.8'
  message_attributes:
    description:
      - Dictionary of message attributes. These are optional structured data entries to be sent along to the endpoint.
      - This is in AWS's distinct Name/Type/Value format; see example below.
  message_structure:
    description:
      - The payload format to use for the message.
      - This must be 'json' to support protocol-specific messages (`http`, `https`, `email`, `sms`, `sqs`). It must be 'string' to support message_attributes.
    default: json
    choices: ['json', 'string']
extends_documentation_fragment:
  - ec2
  - aws
requirements:
  - boto3
  - botocore
"""

EXAMPLES = """
- name: Send default notification message via SNS
  sns:
    msg: '{{ inventory_hostname }} has completed the play.'
    subject: Deploy complete!
    topic: deploy
  delegate_to: localhost

- name: Send notification messages via SNS with short message for SMS
  sns:
    msg: '{{ inventory_hostname }} has completed the play.'
    sms: deployed!
    subject: Deploy complete!
    topic: deploy
  delegate_to: localhost

- name: Send message with message_attributes
  sns:
    topic: "deploy"
    msg: "message with extra details!"
    message_attributes:
      channel:
        data_type: String
        string_value: "mychannel"
      color:
        data_type: String
        string_value: "green"
  delegate_to: localhost
"""

RETURN = """
msg:
  description: Human-readable diagnostic information
  returned: always
  type: string
  sample: OK
message_id:
  description: The message ID of the submitted message
  returned: when success
  type: string
  sample: 2f681ef0-6d76-5c94-99b2-4ae3996ce57b
"""

import json
import traceback

try:
    from botocore.exceptions import BotoCoreError, ClientError
except ImportError:
    pass    # Handled by AnsibleAWSModule

from ansible.module_utils.aws.core import AnsibleAWSModule


def arn_topic_lookup(module, client, short_topic):
    lookup_topic = ':{}'.format(short_topic)

    try:
        paginator = client.get_paginator('list_topics')
        topic_iterator = paginator.paginate()
        for response in topic_iterator:
            for topic in response['Topics']:
                if topic['TopicArn'].endswith(lookup_topic):
                    return topic['TopicArn']
    except (BotoCoreError, ClientError) as e:
        module.fail_json_aws(e, msg='Failed to look up topic ARN')

    return None


def main():
    protocols = [
        'http',
        'https',
        'email',
        'email_json',
        'sms',
        'sqs',
        'application',
        'lambda',
    ]

    argument_spec = dict(
        msg=dict(required=True, aliases=['default']),
        subject=dict(),
        topic=dict(required=True),
        message_attributes=dict(type='dict'),
        message_structure=dict(choices=['json', 'string'], default='json'),
    )

    for p in protocols:
        argument_spec[p] = dict()

    module = AnsibleAWSModule(argument_spec=argument_spec)

    sns_kwargs = dict(
        Message=module.params['msg'],
        Subject=module.params['subject'],
        MessageStructure=module.params['message_structure'],
    )

    if module.params['message_attributes']:
        if module.params['message_structure'] != 'string':
            module.fail_json(msg='message_attributes is only supported when the message_structure is "string".')
        sns_kwargs['MessageAttributes'] = module.params['message_attributes']

    dict_msg = {
        'default': sns_kwargs['Message']
    }

    for p in protocols:
        if module.params[p]:
            if sns_kwargs['MessageStructure'] != 'json':
                module.fail_json(msg='Protocol-specific messages are only supported when message_structure is "json".')
            dict_msg[p.replace('_', '-')] = module.params[p]

    client = module.client('sns')

    topic = module.params['topic']
    if ':' in topic:
        # Short names can't contain ':' so we'll assume this is the full ARN
        sns_kwargs['TopicArn'] = topic
    else:
        sns_kwargs['TopicArn'] = arn_topic_lookup(module, client, topic)

    if not sns_kwargs['TopicArn']:
        module.fail_json(msg='Could not find topic: {}'.format(topic))

    if sns_kwargs['MessageStructure'] == 'json':
        sns_kwargs['Message'] = json.dumps(dict_msg)

    try:
        result = client.publish(**sns_kwargs)
    except (BotoCoreError, ClientError) as e:
        module.fail_json_aws(e, msg='Failed to publish message')

    module.exit_json(msg='OK', message_id=result['MessageId'])


if __name__ == '__main__':
    main()